Blob support

lgrahl opened the pull request:
These changes add `Blob` support. This will break API.

Performance comparison: ~~https://jsperf.com/chunked-dc-js-blob-vs-uint8array~~ (see https://github.com/saltyrtc/chunked-dc-js/pull/21#issuecomment-354850309)
Looks like `Blob` is quite a lot faster than `Uint8Array`.

lgrahl committed:
Add possiblity to chunk/unchunk from Blob-based data

lgrahl committed:
Update dist

dbrgn commented on README.md:
````diff
@@ -79,18 +100,32 @@ The example above will return 3 chunks (header prefix not shown):
+```javascript
+unchunker.onMessage = (message: Blob, context: any[]) => {
+    // Do something with the received message
+};
````
Hm, I just noticed that this is TypeScript syntax, not JavaScript. Not sure if we should reword something...

dbrgn commented on chunked-dc.d.ts:
```diff
@@ -9,35 +9,58 @@ declare namespace chunkedDc {
-    interface ChunkerStatic {
-        new(id: number, message: Uint8Array, chunkSize: number): Chunker
+    interface BlobChunkerStatic {
+        new(id: number, message: Uint8Array, chunkSize: number): BlobChunker
```
That `message` type should probably be a `Blob`

dbrgn commented on src/chunker.ts:
```diff
@@ -10,37 +10,39 @@
  */
-export class Chunker implements chunkedDc.Chunker {
-
+export abstract class Chunker {
```
This should probably not be exported, right?

dbrgn commented on src/chunker.ts:
```diff
@@ -10,37 +10,39 @@
      * @param chunkSize The chunk size *including* header data.
      */
-    constructor(id: number, message: Uint8Array, chunkSize: number) {
+    constructor(id: number, message: any, chunkSize: number) {
```
What about using generics for the `message` type? :)

dbrgn commented on src/unchunker.ts:
```diff
@@ -12,33 +12,25 @@ import {Common} from "./common";
      */
-    constructor(buf: ArrayBuffer, context?: any) {
-        if (buf.byteLength < Common.HEADER_LENGTH) {
+    constructor(buf: any, length: number, context?: any) {
```
Same here, instead of `any` we could use generics.

dbrgn commented on src/unchunker.ts:
```diff
@@ -111,16 +198,19 @@ class ChunkCollector {
-    public merge(): {message: Uint8Array, context: any[]} {
+    public merge(): {message: any, context: any[]} {
+        // TODO: Performance. Merging all at once in the end is usually slow compared
+        //       to merging with each chunk.
```
Is there something that still needs to be done / benchmarked here?

lgrahl replied:
I'd suggest doing that in a separate PR. But yes, this could be quite a performance improvement. I used the same approach in RAWRTC (merging at the end) and it turned out to be significantly slower (a bunch of seconds when merging 1 GiB at once) than merging with each chunk.

dbrgn commented on src/unchunker.ts:
```diff
@@ -98,6 +184,7 @@ class ChunkCollector {
      * Return whether this chunk collector already contains a chunk with the specified serial.
      */
     public hasSerial(serial: number): boolean {
+        // TODO: Performance. Slow?
```
Should this be benchmarked?

lgrahl replied:
I guess we should compare it to a hash map approach (or even combined - array and hash map).

dbrgn commented on src/unchunker.ts:
```diff
@@ -157,51 +305,36 @@ class ChunkCollector {
      * @throws Error if message is smaller than the header length.
      */
-    public add(buf: ArrayBuffer, context?: any): void {
+    public async add(buf: any, context?: any): Promise<void> {
```
:( But I guess it can't be avoided.

dbrgn reviewed (changes requested):
I didn't review everything yet, but left some initial comments.

lgrahl commented:
I actually screwed up the tests. Fixed and split up:
https://jsperf.com/chunked-dc-js-chunking
https://jsperf.com/chunked-dc-js-unchunking

Blob is ridiculously fast in FF 58 when chunking... 3.5 GiB/s vs 28 MiB/s. However, it's pretty slow in Chrome when unchunking... 50 MiB/s vs 240 MiB/s.

lgrahl commented:
I'll close this for now since it's unlikely we'll get `Blob` support for data channels anytime soon. It's more likely that the issue of fragmentation/reassembly will be fixed with a new API. Furthermore, this breaks the API in a way that is very annoying to handle.
